### PR TITLE
Make commands nonblocking

### LINKED ISSUE #321: Long-running commands block the gateway and cause disconnection
a change in this commit removed a hack we had in place to account for this: https://github.com/discord-csharp/MODiX/commit/04846c0fca38e71c631b6e23ca86ad0d3dca5b70#diff-6b1fa8622d02ffc8bbfe04946be2d5d0L210

### PR DESCRIPTION
**Pull request by Scott-Caldwell**
Fixes #321.

This PR changes the default `RunMode` to `RunMode.Async` for all commands.

Scope disposal and error handling have been moved to the new `CommandExecuted` event handler available in Discord.Net 2.0.

### CODE REVIEW

File: Modix.Bot/ModixBot.cs [hunk 1]
```diff
@@ -145,8 +151,9 @@ void OnStopping()
 
                 _client.Log -= _serilogAdapter.HandleLog;
                 _commands.Log -= _serilogAdapter.HandleLog;
-
                 _restClient.Log -= _serilogAdapter.HandleLog;
+
+                _commands.CommandExecuted -= HandleCommandResultAsync;
```
Comment (jrusbatch) [inline Modix.Bot/ModixBot.cs]: Should any remaining command scopes be disposed of here?

Comment (Scott-Caldwell) [reply Modix.Bot/ModixBot.cs]: Good idea

File: Modix.Bot/ModixBot.cs [hunk 2]
```diff
@@ -224,14 +231,23 @@ private async Task HandleCommand(SocketMessage messageParam)
 
             var context = new CommandContext(_client, message);
 
-            using (var scope = _scope.ServiceProvider.CreateScope())
-            {
-                await scope.ServiceProvider
-                    .GetRequiredService<IAuthorizationService>()
-                    .OnAuthenticatedAsync(context.User as IGuildUser);
+            var commandScope = _scope.ServiceProvider.CreateScope();
+            _commandScopes[context] = commandScope;
```
Comment (jrusbatch) [inline Modix.Bot/ModixBot.cs]: Do completed scopes ever get removed?

File: Modix.Bot/ModixBot.cs [hunk 3]
```diff
@@ -224,14 +231,23 @@ private async Task HandleCommand(SocketMessage messageParam)
...
 
-                var result = await _commands.ExecuteAsync(context, argPos, scope.ServiceProvider);
+            await commandScope.ServiceProvider
+                .GetRequiredService<IAuthorizationService>()
+                .OnAuthenticatedAsync(context.User as IGuildUser);
+
+            await _commands.ExecuteAsync(context, argPos, commandScope.ServiceProvider);
```
Comment (jrusbatch) [inline Modix.Bot/ModixBot.cs]: Should the scope for this context be disposed at some point after this line?

File: Modix.Bot/ModixBot.cs [hunk 4]
```diff
@@ -224,14 +231,23 @@ private async Task HandleCommand(SocketMessage messageParam)
...
+
+            stopwatch.Stop();
+            Log.LogInformation($"Took {stopwatch.ElapsedMilliseconds}ms to process: {message}");
+        }
 
+        private async Task HandleCommandResultAsync(Optional<CommandInfo> command, ICommandContext context, IResult result)
+        {
+            using (var commandScope = _commandScopes[context])
```
Comment (jrusbatch) [inline Modix.Bot/ModixBot.cs]: Would it be appropriate to remove the scope from the collection at this point?

File: Modix.Bot/ModixBot.cs [hunk 5]
```diff
@@ -31,7 +32,9 @@ public sealed class ModixBot : BackgroundService
         private readonly ModixConfig _config;
         private readonly DiscordSerilogAdapter _serilogAdapter;
         private readonly IApplicationLifetime _applicationLifetime;
+        private readonly CommandErrorHandler _commandErrorHandler;
         private IServiceScope _scope;
+        private readonly ConcurrentDictionary<ICommandContext, IServiceScope> _commandScopes = new ConcurrentDictionary<ICommandContext, IServiceScope>();
```
Comment (jrusbatch) [inline Modix.Bot/ModixBot.cs]: Do we need to use `ConcurrentDictionary` if we only ever add a brand new instance of `ICommandContext`?

Comment (Scott-Caldwell) [reply Modix.Bot/ModixBot.cs]: Not sure how regular dictionaries handle multithreaded adds of different keys. I thought it wasn't threadsafe and could corrupt the internal collection

File: Modix.Bot/ModixBot.cs (hunk 4 above)
Comment (Scott-Caldwell) [reply Modix.Bot/ModixBot.cs]: Yeah, good idea

File: Modix.Bot/ModixBot.cs (hunk 3 above)
Comment (Scott-Caldwell) [reply Modix.Bot/ModixBot.cs]: It gets disposed after the command finishes, in HandleCommandResultAsync

File: Modix.Bot/ModixBot.cs (hunk 2 above)
Comment (Scott-Caldwell) [reply Modix.Bot/ModixBot.cs]: Woopsies. Fixing.

File: Modix.Bot/ModixBot.cs (hunk 5 above)
Comment (jrusbatch) [reply Modix.Bot/ModixBot.cs]: Sticking with a `ConcurrentDictionary` is fine, it definitely seems like the safer approach.

**Comment by jrusbatch** [comment]
Let's see what happens!

**Pull request merged**